types/errors.go

```diff
@@ -316,3 +320,138 @@ func (e ErrInvalidHeaderDataHashFraud) Error() string {
+	StateIndex uint64
+	SLHeight   uint64
+	DAHeight   uint64
+}
+
+func NewErrStateUpdateHeightNotMatchingFraud(stateIndex uint64, slHeight uint64, daHeight uint64) error {
+	return &ErrStateUpdateHeightNotMatchingFraud{
+		StateIndex: stateIndex,
+		SLHeight:   slHeight,
+		DAHeight:   daHeight,
+	}
+}
+
+func (e ErrStateUpdateHeightNotMatchingFraud) Error() string {
+	return fmt.Sprintf("block height in DA batch not matching SL batch height. StateIndex: %d SLHeight: %d DAHeight: %d", e.StateIndex, e.SLHeight, e.DAHeight)
+}
+
+func (e ErrStateUpdateHeightNotMatchingFraud) Unwrap() error {
+	return gerrc.ErrFault
+}
+
+type ErrStateUpdateStateRootNotMatchingFraud struct {
+	StateIndex  uint64
+	Height      uint64
+	SLStateRoot []byte
+	DAStateRoot []byte
+}
+
+func NewErrStateUpdateStateRootNotMatchingFraud(stateIndex uint64, height uint64, slStateRoot []byte, daStateRoot []byte) error {
+	return &ErrStateUpdateStateRootNotMatchingFraud{
+		StateIndex:  stateIndex,
+		Height:      height,
+		SLStateRoot: slStateRoot,
+		DAStateRoot: daStateRoot,
+	}
+}
+
+func (e ErrStateUpdateStateRootNotMatchingFraud) Error() string {
+	return fmt.Sprintf("state root in DA batch block not matching state root in SL. StateIndex: %d Height: %d SL state root: %s DA state root: %s", e.StateIndex, e.Height, hex.EncodeToString(e.SLStateRoot), hex.EncodeToString(e.DAStateRoot))
+}
+
+func (e ErrStateUpdateStateRootNotMatchingFraud) Unwrap() error {
+	return gerrc.ErrFault
+}
+
+type ErrStateUpdateTimestampNotMatchingFraud struct {
+	StateIndex  uint64
+	Height      uint64
+	SLTimestamp time.Time
+	DATimestamp time.Time
+}
+
+func NewErrStateUpdateTimestampNotMatchingFraud(stateIndex uint64, height uint64, slTimestamp time.Time, daTimestamp time.Time) error {
+	return &ErrStateUpdateTimestampNotMatchingFraud{
+		StateIndex:  stateIndex,
+		Height:      height,
+		SLTimestamp: slTimestamp,
+		DATimestamp: daTimestamp,
+	}
+}
+
+func (e ErrStateUpdateTimestampNotMatchingFraud) Error() string {
+	return fmt.Sprintf("timestamp in DA batch block not matching timestamp in SL. StateIndex: %d Height: %d SL timestsamp: %s DA timestamp: %s", e.StateIndex, e.Height, e.SLTimestamp, e.DATimestamp)
+}
+
+func (e ErrStateUpdateTimestampNotMatchingFraud) Unwrap() error {
+	return gerrc.ErrFault
+}
+
+type ErrStateUpdateDoubleSigningFraud struct {
+	Height uint64
+	Block  []byte
+}
+
+func NewErrStateUpdateDoubleSigningFraud(height uint64, block *Block) error {
+	hash := block.Hash()
+	abciBlock, err := ToABCIBlock(block)
+	if err != nil {
+		return err
+	}
+	resultBlock := &ctypes.ResultBlock{
+		BlockID: tmtypes.BlockID{
+			Hash: hash[:],
+			PartSetHeader: tmtypes.PartSetHeader{
+				Total: 1,
+				Hash:  hash[:],
+			},
+		},
+		Block: abciBlock,
+	}
+
+	jsonBlock, err := legacy.Cdc.MarshalJSON(resultBlock)
+	if err != nil {
+		return err
+	}
+
+	return &ErrStateUpdateDoubleSigningFraud{
+		Height: height,
+		Block:  jsonBlock,
+	}
+}
```

Comment:
obviously not enough to prosecute, but I guess we said that will come in another issue